pkg/services/serviceaccounts/database/database.go

```diff
@@ -61,26 +63,133 @@ func deleteServiceAccountInTransaction(sess *sqlstore.DBSession, orgID, serviceA
 	return nil
 }
 
+// TODO: write test
+/*
+ write benchmarks if possible
+*/
 func (s *ServiceAccountsStoreImpl) UpgradeServiceAccounts(ctx context.Context) error {
+	canStart, err := s.startMigration(ctx, orgID)
+	if err != nil {
+		return err
+	}
+	if !canStart {
+		// inprogress
+		return nil
+	}
 	basicKeys := s.sqlStore.GetNonServiceAccountAPIKeys(ctx)
 	if len(basicKeys) > 0 {
 		s.log.Info("Launching background thread to upgrade API keys to service accounts", "numberKeys", len(basicKeys))
 		go func() {
-			for _, key := range basicKeys {
-				sa, err := s.sqlStore.CreateServiceAccountForApikey(ctx, key.OrgId, key.Name, key.Role)
-				if err != nil {
-					s.log.Error("Failed to create service account for API key", "err", err, "keyId", key.Id)
-					continue
-				}
-
-				err = s.sqlStore.UpdateApikeyServiceAccount(ctx, key.Id, sa.Id)
-				if err != nil {
-					s.log.Error("Failed to attach new service account to API key", "err", err, "keyId", key.Id, "newServiceAccountId", sa.Id)
-					continue
-				}
-				s.log.Debug("Updated basic api key", "keyId", key.Id, "newServiceAccountId", sa.Id)
+			err := s.migrateToServiceAccounts(basicKeys, orgID)
+			if err != nil {
+				s.log.Error("Failed to upgrade API keys to service accounts", "err", err)
 			}
+
 		}()
+	} else {
+		// no basic apikeys found
+		return s.updateMigrationLog(ctx, orgID, true, nil)
 	}
 	return nil
 }
+
+func (s *ServiceAccountsStoreImpl) migrateToServiceAccounts(basicKeys []*models.ApiKey, orgID int64) error {
```

Comment:
You need a mutex here, and probably at the start of every function involved in migrations.